src/filter.ts

```diff
@@ -0,0 +1,357 @@
+        }
+      } else {
+        isPass = pattern.path.value === logPath
+      }
+    }
+
+    // Allow negates to undo previous passes while non-negates can only pass,
+    // not unpass.
+    if (pattern.negate) {
+      yaynay = !isPass
+    } else if (isPass) {
+      yaynay = true
+    }
+  }
+
+  return yaynay
+}
+
+function comp(kind: Parsed['level']['comp'], a: number, b: number): boolean {
+  if (kind === 'eq') return a === b
+  if (kind === 'gte') return a >= b
+  if (kind === 'lte') return a <= b
+  casesHandled(kind)
+}
+
+/**
+ * Like `parse` but throws upon any failure.
+ *
+ * @remarks
+ *
+ * Only use this if you know what you're doing.
+ */
+export function parseUnsafe(defaults: Defaults, pattern: string): Parsed[] {
+  return parse(defaults, pattern).map(rightOrThrow)
+}
+
+type ParseError = ContextualError<{ pattern: string; hint?: string }>
+
+function createInvalidPattern(pattern: string, hint?: string): ParseError {
+  return createContextualError(`Invalid filter pattern: "${pattern}"`, { pattern, hint })
+}
+
+/**
+ * Get the string contents of a manual showing how to write filters.
+ */
+export function renderSyntaxManual() {
+  const m = chalk.magenta
+  const b = chalk.blue
+  const gray = chalk.gray
+  const c = chalk.cyan
+  const bold = chalk.bold
+  const subtle = (x: string) => gray(x)
+  const subtitle = (x: string) => bold(m(x))
+  const pipe = gray('|')
+  return `${bold(b('LOG FILTERING SYNTAX MANUAL  ⟁'))}
+${bold(b(`▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬`))}
+
+${bold(b(`Grammar`))}
+
+    [!](${c(`<path>`)}|*)[@((${c(`<levelNum>`)}|${c(`<levelLabel>`)})[+-]|*)][,<...>]
+
+    ${c(`<path>`)}       = ${validPathSegmentNameRegex.toString().replace(/(^\/|\/$)/g, gray(`$1`))}
+    ${c(`<levelNum>`)}   = 1     ${pipe} 2     ${pipe} 3    ${pipe} 4    ${pipe} 5     ${pipe} 6
+    ${c(`<levelLabel>`)} = trace ${pipe} debug ${pipe} info ${pipe} warn ${pipe} error ${pipe} fatal
+
+${bold(b(`Examples`))}
+
+    ${subtitle(`Paths`)}
+    app         ${subtle(`app path at default level`)}
+    app:router  ${subtle(`app router path at default level`)} 
+
+    ${subtitle(`Wildcards Paths`)}
+    *           ${subtle(`all paths at default level`)} 
+    app:*       ${subtle(`descendent paths of app at defualt level`)}
+    app*        ${subtle(`app path and its descendent paths at defualt level`)}
+
+    ${subtitle(`Negation`)}
+    !app        ${subtle(`all paths expect app at default level`)} 
+
+    ${subtitle(`Lists`)}
+    app,nexus   ${subtle(`app and nexus paths at default level`)}
+
+    ${subtitle(`Levels`)}
+    *@info      ${subtle(`all paths at info level`)}
+    *@error-    ${subtle(`all paths at error level or lower`)}
+    *@debug+    ${subtle(`all paths at debug level or higher`)}
+    *@3         ${subtle(`all paths at info level`)}
+    *@4-        ${subtle(`all paths at error level or lower`)}
+    *@2+        ${subtle(`all paths at debug level or higher`)}
```

Comment:
So, `LOG_FILTER` can be used to set `LOG_LEVEL`, right?

Reply:
Yeah, there's this somewhat awkward balance between defaults and the filter. The filter is the source of truth but when things are left unspecified where should the the value be taken from? In other words `LOG_LEVEL` could be renamed now to `LOG_FILTER_DEFUALT_LEVEL`. But I'm not sure about all this yet.